### PR TITLE
Export warning classes and add PyErr::warn_explicit()

### LINKED ISSUE #2741: export `PyUserWarning`
Just as other exceptions are available from `pyo3::exceptions`.

Perhaps other warnings?

Perhaps from `pyo3::warnings`?

For know, I just need `UserWarning`.

In the meantime, I'll import it manually.

### PR DESCRIPTION
**Pull request by birkenfeld**
Fixes #2741

### CODE REVIEW

**Review by samuelcolvin — COMMENTED** [review]
LGTM, two comments:

* would it be clearer if we moved warnings to `pyo3::warning`?
* How hard would it be to add a method so we could call `PyErr::warn_from_type::<PyUserWarning>(....)` (like you can do `dict.is_instance_of::<PyAny>`)?

**Comment by birkenfeld** [comment]
1. there wouldn't be much in it, and since Warnings are Exceptions anyway...
2. possible, but I feel the combinatoric explosion of methods (what about a counterpart for `warn_explicit`...) is not worth the few times when it is needed.

Of course, these are my opinions, other maintainers may disagree.

File: src/err/mod.rs [hunk 1]
```diff
@@ -492,6 +511,48 @@ impl PyErr {
...
+            Some(s) => s.as_ptr(),
+        };
+        let registry: *mut ffi::PyObject = match registry {
+            None => std::ptr::null_mut(),
+            Some(obj) => obj.as_ptr(),
+        };
+        unsafe {
+            error_on_minusone(
+                py,
+                ffi::PyErr_WarnExplicit(
```
Comment (davidhewitt) [inline src/err/mod.rs:544]: It looks like `PyPyErr_WarnExplicit` does exist as a symbol in PyPy shared lib, so I think we can update the ffi definition and enable `warn_explicit` on pypy.

File: src/err/mod.rs [hunk 2]
```diff
@@ -477,7 +477,26 @@ impl PyErr {
     }
 
     /// Issues a warning message.
```
Comment (davidhewitt) [inline src/err/mod.rs:479]: Perhaps worth expanding this doc to say it's equivalent to `warnings.warn` (and similar for `warn_explicit`)?

**Review by davidhewitt — COMMENTED** [review]
Thanks for this! I'm ok with `pyo3::exceptions` as a reasonable home for these for now, though I'm also not sure either way. It should be possible to re-home them in the future if we want.

As for the typed methods, I'm unconvinced we need them for now - users can always define wrappers e.g. `user_warning(py, msg)` to do exactly what you want.

**Review by davidhewitt — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**